R/bin_spatial.R

```diff
@@ -0,0 +1,220 @@
+#' set to `TRUE`, a list is returned with both the input `occdf` and grid
+#' information.
+#' If the `method` argument is specified as "dist", a list is returned of
+#' spatial samples drawn around unique localities based on the
+#' user-defined `dist`. Each element of the list contains occurrences within
+#' the geographic distance of the reference locality defined by the user.
+#'
+#' @details Two approaches (methods) exist in the `bin_spatial()` function for
+#' assigning occurrences to bins/samples:
+#' - Equal-area grid: The "grid" method bins fossil occurrence data into
+#' equal-area grid cells using discrete hexagonal grids via the
+#' \code{\link[h3jsr]{h3jsr}} package. This package relies on Uber's H3 library,
+#' a geospatial indexing system that partitions the world into hexagonal cells:
+#' \url{https://h3geo.org/docs}. In H3, 16 different resolutions are available:
+#' \url{https://h3geo.org/docs/core-library/restable}. In the implementation of
+#' the `bin_spatial()` function, the resolution is defined by the user-input
+#' `dist` which represents the distance between the centroid of adjacent cells.
+#' Using this distance, the function identifies which resolution is most
+#' similar to the input `dist`, and uses this resolution.
+#' - Distance: The "dist" method identifies unique localities in the input
+#' `occdf` (i.e., unique pairs of coordinates) and generates a spatial sample
+#' for each locality based on a user-defined distance. All occurrences within
+#' the specified distance from the reference locality are drawn as a single
+#' spatial sample. As high density of occurrences might result in numerous
+#' samples for the same area, the `buffer` argument has been implemented to
+#' allow users to generalise the definition of
+#' a unique locality to incorporate a broader geographic area. The
+#' functionality of this approach is heavily
+#' dependent on the
+#' \code{\link[sf:st_is_within_distance]{sf::st_is_within_distance()}} function.
+#'
+#' Note: prior to implementation of either method, the coordinate reference
+#' system (CRS) for input data is defined as EPSG:4326 (World Geodetic System
+#' 1984). The user might wish to update their data accordingly if this is
+#' not appropriate.
+#'
+#' @section Developer(s):
+#' Lewis A. Jones
+#' @section Reviewer(s):
+#' To be reviewed
+#' @importFrom sf st_as_sf st_is_within_distance st_drop_geometry
+#' @importFrom h3jsr point_to_h3 h3_to_point
+#' @examples
+#' # Get internal data
+#' data("tetrapods")
+#' # Smaller dataframe for examples
+#' occdf <- tetrapods[1:500, ]
+#' # Bin data using a hexagonal equal-area grid
+#' bin_spatial(occdf = occdf, method = "grid", dist = 250)
+#' # Bin data based on distance for each unique location
+#' bin_spatial(occdf = occdf, method = "dist", dist = 250)
+#' # Bin data based on distance with a buffer to define each unique location
+#' bin_spatial(occdf = occdf, method = "dist", dist = 250, buffer = 100)
```

Comment:
Examples using the output of palaeorotate should be included when that's ready, and we could also add an example using tax_unique looking for unique taxa within different grid cells :)